### ISSUE TITLE
atexit support

### ISSUE THREAD

**Issue by tomsik68** [staticafi/symbiotic#176]
Looks like sv-comp will include benchmarks using `atexit` function in MemSafety category. 

I tested Symbiotic 3ea77c5 on the testing program below (after preprocessing) and it produces incorrect result. This is the command I used to run Symbiotic: `./symbiotic --sv-comp --prp memcleanup test_atexit.i`. I tried even with `--no-slice` and the problem persists. I expect Symbiotic to report `true`, but it reports `false`. Symbiotic probably doesn't understand atexit even though klee has some support for it.

```c
#include <stdlib.h>

static int *p;

void free_glob(void) {
    free(p);
}

int main() {
    p = (int*) malloc(sizeof(int));
    atexit(free_glob);
    return 0;
}
```

**Comment by mchalupa** [comment]
Yes, the support must include two things:
- [ ] support `atexit` (and `on_exit`?) in KLEE (should be easy to implement)
- [ ] support `atexit` and `on_exit` by slicing. This will probably require setting these functions as the conditional slicing criterina. Hmm, but if the slicing criterion is inside the `atexit` handler, we must treat it specially also in the slicer
- [ ] support in instrumentation. This will probably work out of the box once we have the support in the slicer.

**Comment by tomsik68** [comment]
Perhaps we could support it via instrumentation. Before each atexit call, we would store the passed function pointer into a global array. At the end of main, we would just call all functions from that array. 
This solution is not clean at all, but there is no need to extend slicing nor klee, so that's a plus.

**Comment by tomsik68** [comment]
Using `-libc=klee` fixes the problem, but that implementation doesn't call the registered functions in reverse order as specified by standard.

**Comment by mchalupa** [comment]
> At the end of main, we would just call all functions from that array.

You must do that also on every call of `exit()`, but using instrumentation it is an interesting idea.

> Using -libc=klee fixes the problem, but that implementation doesn't call the registered functions in reverse order as specified by standard.

That that's a bug in KLEE, isn't it?

**Comment by tomsik68** [comment]
Yes, it's bug in KLEE. I will make a pull request which fixes it.

I researched how KLEE handles it. They add a function to `llvm.global_dtors`, which in turn calls everything passed to atexit.
About `global_dtors`:
>The functions referenced by this array will be called in descending order of priority (i.e. highest first) when the module is unloaded.

Question is, do we want to use `-libc=klee` to handle it in Symbiotic? In my opinion, the cleanest solution would be this: 
1. While linking undefined functions, link our own definition of `atexit` which would work exactly like the one in KLEE (https://github.com/staticafi/klee/blob/2805689b338afacf8471527c2a940f483dd5e0ef/runtime/klee-libc/__cxa_atexit.c).
2. Add support for `llvm.global_dtors` to slicer.

**Comment by mchalupa** [comment]
> Question is, do we want to use -libc=klee to handle it in Symbiotic? 

Probably not. IIRC than KLEE moves to using uclibc-klee instead of klee's libc (but uclibc-klee probably handles it the same way).

> Add support for llvm.global_dtors to slicer.

This is a good plan, but it is not enough. We must explicitly handle `atexit` in slicer otherwise it will not work outside of Symbiotic. That is, we should support `global_ctors` and `global_dtors`, but we should also support `atexit` per se.
